Updated dependencies

taneltm opened the pull request:
In addition to updating the dependencies, fixed new linter errors and updated the license year.

Resolves WE2-1048

taneltm committed:
Update dependencies

WE2-1048

Signed-off-by: Tanel Metsar <taneltm@users.noreply.github.com>

taneltm committed:
Fix linter errors

WE2-1048

Signed-off-by: Tanel Metsar <taneltm@users.noreply.github.com>

taneltm committed:
Update copyright year

WE2-1048

Signed-off-by: Tanel Metsar <taneltm@users.noreply.github.com>

mrts commented on rollup.config.mjs:
```diff
@@ -15,7 +15,6 @@ export default {
       banner: {
         content: {
-          // eslint-disable-next-line no-undef
           file:     path.join(__dirname, "LICENSE"),
```
Lets use
```suggestion
          file:     path.join(import.meta.dirname, "LICENSE"),
```
like in the extension project.

taneltm replied:
Thanks for the reminder

taneltm committed:
Use import.meta.dirname instead of __dirname

WE2-1048

Signed-off-by: Tanel Metsar <taneltm@users.noreply.github.com>

mrts commented on eslint.config.mjs:
```diff
@@ -0,0 +1,55 @@
+
+    ignores: [
+      "dist/",
+      "**/__tests__/**/*"
```
Here's a reminder that we want to lint tests as well, but after the linting rules refactor is done.

mrts commented on src/utils/errorSerializer.ts:
```diff
@@ -34,61 +38,64 @@ import UserTimeoutError from "../errors/UserTimeoutError";
+  return (
+    typeof errorObject.name    === "string" &&
+    typeof errorObject.message === "string" &&
+    typeof errorObject.stack   === "string"
```
Do you want to check `errorObject.code` here as well? You are using it in `deserializeError()` line 89 after `isSerializedError()`.

taneltm replied:
Good catch. I was thinking the usual Error object doesn't have the `code` property, but all serialized errors will be custom errors and will have the `code` property.

mrts commented on src/utils/errorSerializer.ts:
```diff
@@ -34,61 +38,64 @@ import UserTimeoutError from "../errors/UserTimeoutError";
-  } = error;
+export function serializeError(error: Error): SerializedError {
+  if (typeof error != "object" || error == null || !(error instanceof Error)) {
+    throw new UnknownError();
```
Should you perhaps return an `UnknownError` here instead of throwing?

taneltm replied:
Yes it should.

mrts commented on src/utils/errorSerializer.ts:
```diff
@@ -34,61 +38,64 @@ import UserTimeoutError from "../errors/UserTimeoutError";
-    stack,
-  } = error;
+export function serializeError(error: Error): SerializedError {
+  if (typeof error != "object" || error == null || !(error instanceof Error)) {
```
Minor, `null` check is redundant as `instanceof` is false for `null`:
```suggestion
  if (typeof error != "object" || !(error instanceof Error)) {
```
![image](https://github.com/user-attachments/assets/2838b1d8-8d8e-4048-9cd6-92167c6f5715)

mrts commented on src/utils/errorSerializer.ts:
```diff
@@ -34,61 +38,64 @@ import UserTimeoutError from "../errors/UserTimeoutError";
+  if (isSerializedError(errorObject)) {
     const CustomError = errorCodeToErrorClass[errorObject.code];
 
     error = new CustomError();
```
Just checking, can every constructor in `errorCodeToErrorClass`  be called with no arguments?

taneltm replied:
Before this PR you couldn't construct some errors without arguments, but those were changed (MissingParameterError and VersionMismatchError) - Making the TypeScript linter rules stricter revealed that.

mrts commented on src/services/WebExtensionService.ts:
```diff
@@ -129,17 +129,17 @@ export default class WebExtensionService {
-  getInitialAction(action: string): string {
-    return action.replace(/-success$|-failure$|-ack$/, "");
+  getInitialAction(action: string): InitialAction {
+    return action.replace(/-success$|-failure$|-ack$/, "") as InitialAction;
```
I would like to see a test for this conversion.

taneltm committed:
Fix errorSerializer fails to include stack trace in Firefox

Includes:
- ESLint is now configured to lint tests in addition to source code
- Added tests for WebExtensionService.getInitialAction
- Fixed deprecated `.toBeCalledTimes` and `.toBeCalledWith` functions in tests
- Fixed linting errors in tests

WE2-1048

Signed-off-by: Tanel Metsar <taneltm@users.noreply.github.com>

mrts commented on src/errors/SerializedError.ts:
```diff
@@ -0,0 +1,32 @@
+  message: string;
+  name?: string;
+  stack?: string;
+  // ...any other fields that might be serialized
```
This comment duplicates the following line, probably should be removed:
```suggestion
```

mrts commented on tsconfig.eslint.json:
```diff
@@ -0,0 +1,17 @@
+{
+  "compilerOptions": {
+    "module": "es6",
+    "target": "es2019",
```
Should the target be es2021, like in eslint.config.mjs? Or even es2024 in both places (see [this](https://github.com/eslint/eslint/pull/17328))?
```suggestion
    "target": "es2021",
```

mrts commented on src/utils/errorSerializer.ts:
```diff
@@ -34,60 +38,81 @@ import UserTimeoutError from "../errors/UserTimeoutError";
     error = new CustomError();
+
+    Object.entries(errorObject).forEach(([key, value]) => {
+      Reflect.set(error, key, value);
```
Same direct assignment question here:
```suggestion
      error[key] = value;
```

mrts commented on src/utils/errorSerializer.ts:
```diff
@@ -34,60 +38,81 @@ import UserTimeoutError from "../errors/UserTimeoutError";
+  const result: Record<string, unknown> = {};
+
+  for (const prop of Object.getOwnPropertyNames(error)) {
+    Reflect.set(result, prop, Reflect.get(error, prop));
```
Wouldn't direct assignment be more readable:
```suggestion
    result[prop] = error[prop];
```
Or is there any specific reason to use `Reflect`?

taneltm replied:
Here `prop` is an unknown string value and can't be used to reference `Error` object's properties - a down side of having strict linter rules.

This would satisfy the linter:
```ts
result[prop] = error[prop as 'name' | 'stack' | 'message'];
```
But this would be misleading as the error object would have other properties and some of them depend on the browser.

Another issue is that if the `error` object is a KnownError, it would have the `code` property which we also want to copy to the serialized result. The following wouldn't work tho:
```ts
result[prop] = error[prop as 'name' | 'stack' | 'message' | 'code'];
```
That's because the `Error` object doesn't have the `code` property.  

Using Reflect, when the property is dynamic, is probably the cleanest way to do it, unless we change the linter rules.

In the following case, the use of reflect isn't actually needed. This is because the `result` is of `SerializedError` type and we can access `result.code` and `result.name` directly.
```ts
  if (!result.code) {
    Reflect.set(result, "code", ErrorCode.ERR_WEBEID_UNKNOWN_ERROR);
    Reflect.set(result, "name", UnknownError.name);
  }
```
I'll change that.

mrts commented on src/utils/errorSerializer.ts:
```diff
@@ -34,60 +38,81 @@ import UserTimeoutError from "../errors/UserTimeoutError";
+  for (const key of ["name", "stack"]) {
+    const errorAsObject = (error as unknown as Record<string, unknown>);
+    if (!result[key] && errorAsObject?.[key] != null) {
+      Reflect.set(result, key, errorAsObject[key]);
```
Same direct assignment question here.

mrts commented on src/utils/errorSerializer.ts:
```diff
@@ -34,60 +38,81 @@ import UserTimeoutError from "../errors/UserTimeoutError";
-    stack,
-  };
+export function serializeError(error: Error): SerializedError {
+  if (typeof error != "object" || !(error instanceof Error)) {
```
The `typeof` check looks redundant as no non-object can be an instance of `Error`:
```suggestion
  if (!(error instanceof Error)) {
```

taneltm committed:
Fix errors while trying to import the library to an Angular project

WE2-1048

Signed-off-by: Tanel Metsar <taneltm@users.noreply.github.com>
